### DISCUSSION TITLE
Embed AI parameters (model, prompt..) in Exif Metadata?

### DISCUSSION THREAD

**Discussion by daniel-hmc** [lllyasviel/Fooocus#2236] [General]
Hello :) 

Is there a known way to let Fooocus write the image generation parameters (such as used models, prompt, seed...) into the metadata header of the output files, so that e.g. Adobe Lightroom can read an use them? 
PNG even supports Exif metadata since 2017, but it may be safer to let the generated image automatically be converted to JPEG and then let Fooocus add the metadata to it. 

Would be quite useful to have the parameters directly attached to the image instead of only in the log. 

Thanks
Daniel

**Comment by mashb1t** [comment]
see https://github.com/lllyasviel/Fooocus/pull/1940#issuecomment-1931742982

**Reply by poor7** [reply]
Is your implementation optional? Those. can you turn it on and off?

**Reply by mashb1t** [reply]
yes, can be turned on/off in args, config and UI

**Comment by daniel-hmc** [comment]
Cool, thank you. Strange I didn't find that when searching before. 
Sorry, I'm not a github expert. Can you give me a short hint how I can get this into my Fooocus instance please?

**Reply by mashb1t** [reply]
apply the changes manually, use my fork https://github.com/mashb1t/Fooocus or wait until the PR is merged

**Comment by daniel-hmc** [comment]
Thank you, I used your fork. I have enabled metadata embedding and set format to a1111, because I like to have it well-readable.
In a hex / text viewer, I can see the data being embedded in the jpg file.
Questions:
1. I don't see the embedded data anywhere in any other software (Lightroom, Path Finder "Info" panel etc.). How / where is it embedded? What field do I have to look for? All I can see is the "Software" field, which says "Fooocus 2.1.864". In the hex / text viewer, after this there are just two 0x00 bytes, then the data is visible. 
2. The UI often shows "Error" and isn't updated anymore after a hew renderings. I haven't had this on the original version of Fooocus. Known issue / can I avoid this?

![image](https://github.com/lllyasviel/Fooocus/assets/25068087/bff89293-1d7c-4b93-a368-aeb190a67957)

**Reply by mashb1t** [reply]
@daniel-hmc 
1. you have to look for the user comment field, see https://github.com/mashb1t/Fooocus/blob/main/modules/meta_parser.py#L557 and below (same scheme as A1111).
2. i am neither aware nor have encountered disconnect issues in my fork. Please make sure your system resources are not maxed and check + provide terminal output in case of errors in am issue in my fork. Thanks!

**Reply by daniel-hmc** [reply]
@mashb1t regarding 2:
I have the GUI "Error" regularly now, and in terminal output I always see this kind of error then (also copied some context).
Resources are by far not maxed out, as far as I can tell. 

```

[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 6.25 seconds
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 60/60 [03:58<00:00,  3.98s/it]
Requested to load AutoencoderKL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 0.60 seconds
Image generated with private log at: /Users/daniel/fooocus/Fooocus/outputs/2024-02-16/log.html
Generating and saving time: 247.90 seconds
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
**Loading 1 new model
Traceback (most recent call last):
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/routes.py", line 488, in run_predict
    output = await app.get_blocks().process_api(
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/blocks.py", line 1434, in process_api
    data = self.postprocess_data(fn_index, result["prediction"], state)
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/blocks.py", line 1326, in postprocess_data
    prediction_value = postprocess_update_dict(
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/blocks.py", line 460, in postprocess_update_dict
    prediction_value["value"] = block.postprocess(prediction_value["value"])
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/components/gallery.py", line 203, in postprocess
    file_path = self.make_temp_copy_if_needed(img)
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/components/base.py", line 226, in make_temp_copy_if_needed
    temp_dir = self.hash_file(file_path)
  File "/Users/daniel/miniconda3/envs/fooocus/lib/python3.10/site-packages/gradio/components/base.py", line 190, in hash_file
    with open(file_path, "rb") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/daniel/fooocus/Fooocus/outputs/2024-02-16/2024-02-16_20-21-56_1977.jpg'**
[Fooocus Model Management] Moving model(s) has taken 4.77 seconds
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 60/60 [04:07<00:00,  4.13s/it]
Requested to load AutoencoderKL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 0.59 seconds
Image generated with private log at: /Users/daniel/fooocus/Fooocus/outputs/2024-02-16/log.html
Generating and saving time: 255.22 seconds
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 3.10 seconds
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 60/60 [04:02<00:00,  4.04s/it]
Requested to load AutoencoderKL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 0.59 seconds
Image generated with private log at: /Users/daniel/fooocus/Fooocus/outputs/2024-02-16/log.html
Generating and saving time: 247.99 seconds
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
Loading 1 new model
[

```

**Reply by daniel-hmc** [reply]
@mashb1t - I have an idea now. This "**FileNotFoundError**" for the output file leads me into thinking that maybe my Lightoom Auto-Import is causing this issue, since it removes the output file from its original location and puts it into a Lightoom-maintained folder. 
**Can deleting the output file directly after it has been generated cause this GUI Error?**

**Reply by mashb1t** [reply]
@daniel-hmc this is indeed highly possible!

**Reply by daniel-hmc** [reply]
@mashb1t I changed my workflow, deactivated this automatic image movement (effectively: deletion from the "output" directory) and don't get any GUI Error anymore.

**Comment by daniel-hmc** [comment]
Thank you!
1. I have, but the field is empty. See screenshots below.
2. okay, will observe. I let it run over night now and "Error" did not happen. Could have been a performance issue indeed.

![image](https://github.com/lllyasviel/Fooocus/assets/25068087/df26a36c-93b3-47a2-a45a-22d4f203f832)

![image](https://github.com/lllyasviel/Fooocus/assets/25068087/7b33b5a3-47a6-4251-9f47-60af8587c888)

**Comment by daniel-hmc** [comment]
Addition:
If I choose the "fooocus" metadata scheme, that is visible in the User Comment field. But the a1111 scheme metadata isn't. At least not in Lightroom Classic. 
Maybe due to certain characters that are part of the string you fill in there? 
I can do some more tests to find out later. 
Do you know of any Mac metadata editor that CAN display the a1111 scheme metadata that I could use for the tests?

**Reply by mashb1t** [reply]
a1111 uses multiline encoded text while fooocus is one line of json string, maybe that's the issue with Lightroom.
I don't, i've used a free online EXIF checker to ensure correct format as well as python to read again from file to UI.
